### PR TITLE
JumpStart Gated Model Support in ModelBuilder Local Modes

### PR DESCRIPTION
**Pull request by makungaj1**
*Issue #, if available:*

*Description of changes:*
JumpStart does not give customers direct access to gated model weights, this means that ModelBuilder's local modes cannot operate with a JumpStart gated models as weights need to be downloaded in customers' device.

This PR restricts JumpStart Gated Model to only ModelBuilder's endpoint mode. (Backward compatible)

* Example code to deploy a JumpStart gated model on a SM endpoint with ModelBuilder.
```
model_builder = ModelBuilder(
    model=js_gated_model_id,
    schema_builder=SchemaBuilder(sample_input, sample_output),
    mode=Mode.SAGEMAKER_ENDPOINT
)

model = model_builder.build()

predictor = model.deploy(accept_eula=True)
```

*Testing done:*
* Unit Test
* Integration test
* E2E with Notebook

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integ/sagemaker/serve/test_serve_js_happy.py [hunk 1]
```diff
@@ -46,6 +48,17 @@ def happy_model_builder(sagemaker_session):
     )
 
 
+@pytest.fixture
+def happy_model_builder_gated_model(sagemaker_session):
```
Comment (samruds) [inline tests/integ/sagemaker/serve/test_serve_js_happy.py:52]: Can we add a test for a non jumpstart model + has keyword private + wants to test in local mode? We should let such a test case pass through.

**Review by samruds — CHANGES_REQUESTED** [review]

**Review by samruds — APPROVED** [review]

**Review by gwang111 — APPROVED** [review]

File: tests/integ/sagemaker/serve/test_serve_js_happy.py (hunk 1 above)
Comment (makungaj1) [reply tests/integ/sagemaker/serve/test_serve_js_happy.py:52]: Done

File: src/sagemaker/serve/builder/jumpstart_builder.py [hunk 2]
```diff
@@ -469,3 +474,10 @@ def _build_for_jumpstart(self):
             )
 
         return self.pysdk_model
+
+    def _is_gated_model(self, model) -> bool:
+        """Determine if ``this`` Model is Gated"""
```
Comment (jiapinw) [inline src/sagemaker/serve/builder/jumpstart_builder.py:479]: nit: can we start having docstrings as the following from now on:
    """Checks if the provided model is gated.

    Args:
        model: The model object.
    Returns:
        bool: True if is gated model.
    """
    
   You might also want to have some type hints, ie. model: Model

**Review by jiapinw — APPROVED** [review]

File: src/sagemaker/serve/builder/jumpstart_builder.py [hunk 3]
```diff
@@ -469,3 +474,16 @@ def _build_for_jumpstart(self):
...
+        """Determine if ``this`` Model is Gated
+
+        Args:
+            model (Model): Jumpstart Model
+        Returns:
+            bool: ``True`` if ``this`` Model is Gated
+        """
+        s3_uri = model.model_data
+        if isinstance(s3_uri, dict):
+            s3_uri = s3_uri.get("S3DataSource").get("S3Uri")
```
Comment (mufaddal-rohawala) [inline src/sagemaker/serve/builder/jumpstart_builder.py:488]: what if this gets evaluated to None?
```
>>> "private" in None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: argument of type 'NoneType' is not iterable
```

**Review by mufaddal-rohawala — APPROVED** [review]

**Pull request merged**